src/cpp/session/modules/rmarkdown/SessionExecuteChunkOperation.hpp

```diff
@@ -0,0 +1,308 @@
+      Error error = Success();
+      
+      // ensure staging directory
+      FilePath stagingPath = chunkOutputPath(
+               docId_,
+               chunkId_ + kStagingSuffix,
+               ContextExact);
+      
+      error = stagingPath.removeIfExists();
+      if (error)
+         LOG_ERROR(error);
+      
+      error = stagingPath.ensureDirectory();
+      if (error)
+         LOG_ERROR(error);
+      
+      // ensure regular directory
+      FilePath outputPath = chunkOutputPath(
+               docId_,
+               chunkId_,
+               ContextExact);
+      
+      error = outputPath.removeIfExists();
+      if (error)
+         LOG_ERROR(error);
+      
+      error = outputPath.ensureDirectory();
+      if (error)
+         LOG_ERROR(error);
+      
+      // clean old chunk output
+      error = cleanChunkOutput(docId_, chunkId_, true);
+      if (error)
+         LOG_ERROR(error);
+   }
+   
+public:
+   
+   ProcessCallbacks processCallbacks()
+   {
+      ProcessCallbacks callbacks;
+      
+      callbacks.onStarted = boost::bind(
+               &ExecuteChunkOperation::onStarted,
+               shared_from_this(),
+               _1);
+      
+      callbacks.onContinue = boost::bind(
+               &ExecuteChunkOperation::onContinue,
+               shared_from_this(),
+               _1);
+      
+      callbacks.onStdout = boost::bind(
+               &ExecuteChunkOperation::onStdout,
+               shared_from_this(),
+               _1,
+               _2);
+      
+      callbacks.onStderr = boost::bind(
+               &ExecuteChunkOperation::onStderr,
+               shared_from_this(),
+               _1,
+               _2);
+      
+      callbacks.onExit = boost::bind(
+               &ExecuteChunkOperation::onExit,
+               shared_from_this(),
+               _1);
+      
+      return callbacks;
+   }
+   
+private:
+   
+   enum OutputType { OUTPUT_STDOUT, OUTPUT_STDERR };
+   
+   void onStarted(ProcessOperations& operations)
+   {
+      isRunning_ = true;
+   }
+   
+   bool onContinue(ProcessOperations& operations)
+   {
+      return !terminationRequested_;
+   }
+   
+   void onExit(int exitStatus)
+   {
+      events().onChunkExecCompleted(docId_, chunkId_, notebookCtxId());
+      deregisterProcess();
+      isRunning_ = false;
+   }
+   
+   void onStdout(ProcessOperations& operations, const std::string& output)
+   {
+      onText(output, OUTPUT_STDOUT);
+   }
+   
+   void onStderr(ProcessOperations& operations, const std::string& output)
+   {
+      onText(output, OUTPUT_STDERR);
+   }
+   
+   void onText(const std::string& output, OutputType outputType)
+   {
+      using namespace core;
+      
+      // get path to cache file
+      FilePath target = chunkOutputFile(docId_, chunkId_, kChunkOutputText);
+      
+      // generate CSV to write
+      std::vector<std::string> values;
+      values.push_back(outputType == OUTPUT_STDOUT ? "1" : "2");
```

Comment:
For consistency these should use constants:

https://github.com/rstudio/rstudio/blob/3d142e855fb40be8f25341b968f2947c8b75c600/src/cpp/session/modules/rmarkdown/NotebookOutput.hpp#L31-L32